### PR TITLE
Use Source Generators for SvgElementFactory

### PR DESCRIPTION
**Pull request by wieslawsoltes**
#### Reference Issue

Fixes: #771
Fixes: #776
Fixes partially #767 (part 2 of the issue)

#### What does this implement/fix? Explain your changes.

* Make SvgElementFactory faster by removing unnecessary runtime reflection with `Source Generators` as they are compile time operation.
* Removes `Activator.CreateInstance` usage in favor of `new` factory.
* No Source Generators support for `net452` , added additional target framework `net461` because it's min version for source generators.
* This should make code more AOT compilation friendly.

#### Any other comments?

This also makes similar changes as in #764, but initialization is done during compile time when using Source Generators instead in the static ctor.

#### Candidates for optimization using Source Generators

- [x] PropertyDescriptor _svgPropertyAttributes
- [x] PropertyDescriptor _svgEventAttributes 
https://github.com/vvvv/SVG/blob/66a72e474e5efd67b0bf3209c73b228063320b70/Source/SvgElement.cs#L539-L547

- [x] SetPropertyValue
https://github.com/vvvv/SVG/blob/66a72e474e5efd67b0bf3209c73b228063320b70/Source/SvgElementFactory.cs#L174

- [x] SetPropertyValue
https://github.com/vvvv/SVG/blob/66a72e474e5efd67b0bf3209c73b228063320b70/Source/SvgElementFactory.cs#L255

- [x] PropertyDescriptor ElementName
https://github.com/vvvv/SVG/blob/49f52e9073b0638db4c5daefc0ed569498b277d0/Source/SvgElement.cs#L132

### CODE REVIEW

**Comment by wieslawsoltes** [comment]
@inforithmics Can you test this PR if it works with UWP and if it improves performance. This should be now usable, I think I can do some more work to improve perf.

**Comment by wieslawsoltes** [comment]
This is ready for review, I plan to work on more improvements but they can be done after this gets merged.

**Comment by wieslawsoltes** [comment]
The generated code can be found: 

`SVG\Source\obj\GeneratedFiles\Svg.Generators\Svg.Generators.AvailableElementsGenerator\`

File: Generators/AvailableElementsGenerator.cs
Comment (mrbean-bremen) [inline Generators/AvailableElementsGenerator.cs:29]: can be removed

**Review by mrbean-bremen — COMMENTED** [review]
Looks good to me, though I'm not sure about .NET 4.6.1 support (and .NET 4.x support generally) here. You need C# 9 for the source generators to work (also you use some of the language features, though this could be changed), and while strictly speaking the C# version can be different from the .NET version, usually C# 6 comes with 4.6.1 (and C# 7/8 with 4.7 / 4.8).

File: .gitignore
Comment (wieslawsoltes) [reply .gitignore:355]: I am using Rider for past few moths exclusively.

File: Source/Svg.csproj
Comment (wieslawsoltes) [reply Source/Svg.csproj:72]: Source Generators are using netstandard2.0 and this is available in net461, this is just necessary for compilation.

File: Generators/AvailableElementsGenerator.cs
Comment (wieslawsoltes) [reply Generators/AvailableElementsGenerator.cs:29]: Necessary for Source Generators debugging. When you uncomment during build debugger will be attached, this is recommend way to debug.

File: Generators/AvailableElementsGenerator.cs
Comment (wieslawsoltes) [reply Generators/AvailableElementsGenerator.cs:10]: I will add some docs.

File: Samples/SVGViewer/SVGViewer.csproj
Comment (wieslawsoltes) [reply Samples/SVGViewer/SVGViewer.csproj:13]: The net452 does not support source generators, so better for testing.

File: Tests/SvgW3CTestRunner/SvgW3CTestRunner.csproj
Comment (wieslawsoltes) [reply Tests/SvgW3CTestRunner/SvgW3CTestRunner.csproj:13]: The net452 does not support source generators, so better for testing.

File: Generators/AvailableElementsGenerator.cs
Comment (wieslawsoltes) [reply Generators/AvailableElementsGenerator.cs:10]: Added documentation

**Comment by wieslawsoltes** [comment]
> Looks good to me, though I'm not sure about .NET 4.6.1 support (and .NET 4.x support generally) here. You need C# 9 for the source generators to work (also you use some of the language features, though this could be changed), and while strictly speaking the C# version can be different from the .NET version, usually C# 6 comes with 4.6.1 (and C# 7/8 with 4.7 / 4.8).

I have changed the generated code to not include new language features.

I think having net461 support is good as it will benefit from perf optimizations provided from generated code.

For compilation you need new tooling and C# 9, but for end used you don't if consumed via nuget.

**Review by mrbean-bremen — APPROVED** [review]
Ok, that looks good! I will wait for @inforithmics to have a look and check how it works with UWP before merging.

**Comment by inforithmics** [comment]
I have looked at it now. Congrats amazing work. 
1. UWP works without problems. (The less Reflection the better, because UWP is by default AOT compiled)
2. I made a fast Performance Profile.

Current Performance Profile (without any changes)

image

After Performance changes

image

So performance improved from 5 sec to 1.9 sec.

**Comment by inforithmics** [comment]
I had a quick view why the creation of an SvgElement takes so much time and I found some more reflection. On every instantiation of an svg element (_svgPropertyAttributes, _svgEventAttributes)

image

**Comment by inforithmics** [comment]
The other performance bottleneck is parsing of svg Pathes. 

image

Here the problems seems to be creating of substrings, like here instead using for example a Span.

 result = float.Parse(_coords.Substring(Position, i - Position), NumberStyles.Float, CultureInfo.InvariantCulture);

**Comment by mrbean-bremen** [comment]
Thanks - that is certainly something that shall be looked into!
@wieslawsoltes - I will merge as soon as the build is fixed, the last commit broke it. Do you think the other reflection issue found by @inforithmics can also be sped up using source generation?

**Comment by wieslawsoltes** [comment]
> I had a quick view why the creation of an SvgElement takes so much time and I found some more reflection. On every instantiation of an svg element (_svgPropertyAttributes, _svgEventAttributes)
> 
> image

I will fix this.

**Comment by wieslawsoltes** [comment]
> Thanks - that is certainly something that shall be looked into!
> @wieslawsoltes - I will merge as soon as the build is fixed, the last commit broke it. Do you think the other reflection issue found by @inforithmics can also be sped up using source generation?

@mrbean-bremen I am working on reflection issues fix. Please do not merge this PR yet!

**Comment by wieslawsoltes** [comment]
> The other performance bottleneck is parsing of svg Pathes.
> 
> image
> 
> Here the problems seems to be creating of substrings, like here instead using for example a Span.
> 
> result = float.Parse(_coords.Substring(Position, i - Position), NumberStyles.Float, CultureInfo.InvariantCulture);

Yes I saw that in profiler. I need to finish reflection issues first, that will make huge impact on performance. I have already made a lot of progress in that regard. Will push changes to this pr as they can use same source generator.

**Comment by wieslawsoltes** [comment]
> The other performance bottleneck is parsing of svg Pathes.
> 
> image
> 
> Here the problems seems to be creating of substrings, like here instead using for example a Span.
> 
> result = float.Parse(_coords.Substring(Position, i - Position), NumberStyles.Float, CultureInfo.InvariantCulture);

@inforithmics Are you able to look at this and provide solution with e.g. Span ?

**Comment by mrbean-bremen** [comment]
> Please do not merge this PR yet!

Sure, I'll wait until you give the go.

**Comment by inforithmics** [comment]
> > The other performance bottleneck is parsing of svg Pathes.
> > image
> > Here the problems seems to be creating of substrings, like here instead using for example a Span.
> > result = float.Parse(_coords.Substring(Position, i - Position), NumberStyles.Float, CultureInfo.InvariantCulture);
> 
> @inforithmics Are you able to look at this and provide solution with e.g. Span ?

I will investigate performance improvements in this area. If I find something useful I will create a pull request.

**Comment by wieslawsoltes** [comment]
> > > The other performance bottleneck is parsing of svg Pathes.
> > > image
> > > Here the problems seems to be creating of substrings, like here instead using for example a Span.
> > > result = float.Parse(_coords.Substring(Position, i - Position), NumberStyles.Float, CultureInfo.InvariantCulture);
> > 
> > 
> > @inforithmics Are you able to look at this and provide solution with e.g. Span ?
> 
> I will investigate performance improvements in this area. If I find something useful I will create a pull request.

@inforithmics I have also started some work in this area https://github.com/vvvv/SVG/pull/780

**Comment by wieslawsoltes** [comment]
@inforithmics @mrbean-bremen The setting and getting properties without reflection is implemented and ready for review (all unit tests pass), only writing event has been left to implement (not even sure if someone is using this feature):

https://github.com/vvvv/SVG/pull/772/files#diff-9ed8f7a9ce9a87073385867e184b023f7abc478a1b95dfbceb2ad77df7c01cacR604

https://github.com/vvvv/SVG/pull/772/files#diff-9ed8f7a9ce9a87073385867e184b023f7abc478a1b95dfbceb2ad77df7c01cacR1145

**Comment by mrbean-bremen** [comment]
@wieslawsoltes - I'll probably have a closer look only tomorrow evening. My brain is not working well at the moment, as I'm a bit tired today (first work day after the holidays...). If @inforithmics will review the changes, that would be sufficient for me.

**Comment by wieslawsoltes** [comment]
> @wieslawsoltes - I'll probably have a closer look only tomorrow evening. My brain is not working well at the moment, as I'm a bit tired today (first work day after the holidays...). If @inforithmics will review the changes, that would be sufficient for me.

Not problem, I have found few more places to optimize and have few TODO items to finish in this PR anyway.

File: Source/SvgElement.cs
Comment (inforithmics) [inline Source/SvgElement.cs:137]: Shouldn't this special case be added in the Generator?.
So that this Code Path stays

 if (SvgElements.ElementNames.TryGetValue(this.GetType(), out var elementName))
{
    this._elementName = elementName;
}

Comment (wieslawsoltes) [reply Source/SvgElement.cs:137]: SvgDocument does not have SvgElement attribute set, instead the attitude is used on SvgFragment so there would be duplicate im dictionary. I think the way is SvgDocument used this is better place for that exception to attribute convention used.

Comment (inforithmics) [reply Source/SvgElement.cs:137]: That makes sense. So I agree that this is the better solution

Comment (wieslawsoltes) [reply Source/SvgElement.cs:138]: Added comment.

**Comment by inforithmics** [comment]
I made again a measurement with the latest changes. Create Element takes now 1.0 sec instead of 1.9s in the last measuerment

image

Keep up the great work

**Review by inforithmics — APPROVED** [review]

**Comment by wieslawsoltes** [comment]
> I made again a measurement with the latest changes. Create Element takes now 1.0 sec instead of 1.9s in the last measurement
> 
> image
> 
> Keep up the great work.

It would be interesting to compare performance between this PR to latest release.

**Comment by inforithmics** [comment]
> > I made again a measurement with the latest changes. Create Element takes now 1.0 sec instead of 1.9s in the last measurement
> > image
> > Keep up the great work.
> 
> It would be interesting to compare performance between this PR and the latest release.

https://github.com/vvvv/SVG/pull/772#issuecomment-753594215 the first profile so the Create Element went from 5 sec to 1 sec.

**Comment by wieslawsoltes** [comment]
I did run some benachmarks (on VM so result are not 100% good, will run later or real PC).

* tag/nuget-3.1.1


```
BenchmarkDotNet=v0.12.1, OS=ubuntu 20.04
Intel Xeon Platinum 8272CL CPU 2.60GHz, 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=5.0.100
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  Job-EQSVEF : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT

Runtime=.NET Core 3.1  

|               Method |         Mean |      Error |     StdDev |      Op/s | Rank |     Gen 0 |    Gen 1 | Gen 2 |   Allocated |
|--------------------- |-------------:|-----------:|-----------:|----------:|-----:|----------:|---------:|------:|------------:|
| SvgPathBuilder_Parse |     79.78 us |   0.412 us |   0.365 us | 12,534.30 |    1 |    1.5869 |        - |     - |    30.45 KB |
|    AJ_Digital_Camera | 15,299.85 us |  73.293 us |  64.972 us |     65.36 |    2 |  296.8750 | 140.6250 |     - |  5614.54 KB |
|         Issue_134_01 | 71,937.67 us | 711.739 us | 555.679 us |     13.90 |    4 | 1285.7143 | 571.4286 |     - | 23630.96 KB |
|                Tiger | 16,539.72 us |  75.853 us |  70.953 us |     60.46 |    3 |  312.5000 | 125.0000 |     - |  5963.39 KB |

```

* Generators


```
BenchmarkDotNet=v0.12.1, OS=ubuntu 20.04
Intel Xeon Platinum 8272CL CPU 2.60GHz, 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=5.0.100
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  Job-EYUCUU : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT

Runtime=.NET Core 3.1  

|               Method |         Mean |      Error |     StdDev |      Op/s | Rank |  Gen 0 | Gen 1 | Gen 2 |   Allocated |
|--------------------- |-------------:|-----------:|-----------:|----------:|-----:|-------:|------:|------:|------------:|
| SvgPathBuilder_Parse |     63.16 us |   0.255 us |   0.226 us | 15,832.86 |    1 | 0.3662 |     - |     - |     7.17 KB |
|    AJ_Digital_Camera | 10,673.57 us | 166.568 us | 155.808 us |     93.69 |    3 |      - |     - |     - |  3056.59 KB |
|         Issue_134_01 | 28,456.64 us | 550.772 us | 964.633 us |     35.14 |    4 |      - |     - |     - | 12459.99 KB |
|                Tiger |  9,053.13 us | 153.647 us | 143.721 us |    110.46 |    2 |      - |     - |     - |  2327.02 KB |
```


* master


```
BenchmarkDotNet=v0.12.1, OS=ubuntu 20.04
Intel Xeon Platinum 8272CL CPU 2.60GHz, 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=5.0.100
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  Job-ZVYPGF : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT

Runtime=.NET Core 3.1  

|               Method |         Mean |      Error |     StdDev |      Op/s | Rank |    Gen 0 |    Gen 1 | Gen 2 |   Allocated |
|--------------------- |-------------:|-----------:|-----------:|----------:|-----:|---------:|---------:|------:|------------:|
| SvgPathBuilder_Parse |     60.99 us |   0.263 us |   0.220 us | 16,395.77 |    1 |   0.3662 |        - |     - |     7.17 KB |
|    AJ_Digital_Camera | 12,039.46 us |  50.451 us |  42.129 us |     83.06 |    3 | 187.5000 |  78.1250 |     - |  3644.73 KB |
|         Issue_134_01 | 55,952.33 us | 621.247 us | 581.114 us |     17.87 |    4 | 800.0000 | 400.0000 |     - | 14630.31 KB |
|                Tiger | 11,884.51 us |  35.573 us |  33.275 us |     84.14 |    2 | 156.2500 |  78.1250 |     - |  2917.51 KB |

```

**Comment by mrbean-bremen** [comment]
@wieslawsoltes - is this ready to merge? Most of it has been reviewed, are there any open issues you want to do first?
I think I will merge the 3 small PRs first, they shall not have any impact on this, and merge this one if you give the go. It has been a lot of nice work, and from here it really looks ready for integration.

**Comment by wieslawsoltes** [comment]
@mrbean-bremen I will try to finish _svgEventAttributes generator today, this needs field setters and getter for events, similar as for properties.

**Comment by mrbean-bremen** [comment]
Ok, let me me know when you are ready. At your pace, that will not be long... :)

**Comment by wieslawsoltes** [comment]
> Ok, let me me know when you are ready. At your pace, that will not be long... :)

Yeah had to refactor code generation as for event attributes we need spacial handling, and in process I have found better way to generate property descriptors. Will need bit more time to polish the generated code and fix event setters.

**Comment by wieslawsoltes** [comment]
@mrbean-bremen This is ready for final review and merge.

@inforithmics Can you test again this PR, I think it ready.

**Comment by inforithmics** [comment]
Amazing Work. The Create Element takes now 0.5 sec so again the time was halfed after last Performance Profile

image

With the Dtd Optimization the overlad load now takes only 2.3 sec instead of 13.7 sec so overall more than five times faster.

And UWP still works.

**Comment by inforithmics** [comment]
I had a quick look at the generated source code 

```
// <auto-generated />
#nullable disable
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Globalization;

namespace Svg
{
    [AttributeUsage(AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
    public sealed class ElementFactoryAttribute : Attribute
    {
    }

    internal enum DescriptorType
    {
        Property,
        Event
    }

    internal interface ISvgPropertyDescriptor
    {
        DescriptorType DescriptorType { get; }
        string AttributeName { get; }
        string AttributeNamespace { get; }
        TypeConverter Converter { get; }
        Type Type { get; }
        object GetValue(object component);
        void SetValue(object component, ITypeDescriptorContext context, CultureInfo culture, object value);
    }

    internal class SvgPropertyDescriptor<T, TU> : ISvgPropertyDescriptor
    {
        public DescriptorType DescriptorType { get; }
        public string AttributeName { get; }
        public string AttributeNamespace { get; }
        public TypeConverter Converter { get; }
        public Type Type { get; } = typeof(TU);
        private Func<T, TU> Getter { get; }
        private Action<T, TU> Setter { get; }

        public SvgPropertyDescriptor(DescriptorType descriptorType, string attributeName, string attributeNamespace, TypeConverter converter, Func<T, TU> getter, Action<T, TU> setter)
        {
            DescriptorType = descriptorType;
            AttributeName = attributeName;
            AttributeNamespace = attributeNamespace;
            Converter = converter;
            Getter = getter;
            Setter = setter;
        }

        public object GetValue(object component)
        {
            return (object)Getter((T)component);
        }

        public void SetValue(object component, ITypeDescriptorContext context, CultureInfo culture, object value)
        {
            if (Converter != null)
            {
                Setter((T)component, (TU)Converter.ConvertFrom(context, culture, value));
            }
        }
    }
}
```

On the Set Value the value is only set if the Converter is not null, 
So my 
1. question is: does every Property have a Converter.
2. question is: Shouldn't it be a cast if there is no Converter?

```
        public void SetValue(object component, ITypeDescriptorContext context, CultureInfo culture, object value)
        {
            if (Converter != null)
            {
                Setter((T)component, (TU)Converter.ConvertFrom(context, culture, value));
            }
            else
            {
                Setter((T)component, (TU)value);
            }
        }
```

**Comment by wieslawsoltes** [comment]
> Amazing Work. The Create Element takes now 0.5 sec so again the time was halfed after last Performance Profile
> 
> With the Dtd Optimization the overall load now takes only 2.3 sec instead of 13.7 sec so overall more than five times faster.
> 
> And UWP still works.

That is good to hear, I was a bit worried as I had to change how I store property setter (on each class instead of globally) and had to use virtual methods for setters/getters (it was necessary for correctness and event getters), but turns out it is faster solution after all.

Good to hear UWP works.

**Comment by wieslawsoltes** [comment]
> On the Set Value the value is only set if the Converter is not null,
> So my
> 
> 1. question is: does every Property have a Converter.
> 2. question is: Shouldn't it be a cast if there is no Converter?
> 
> ```
>         public void SetValue(object component, ITypeDescriptorContext context, CultureInfo culture, object value)
>         {
>             if (Converter != null)
>             {
>                 Setter((T)component, (TU)Converter.ConvertFrom(context, culture, value));
>             }
>             else
>             {
>                 Setter((T)component, (TU)value);
>             }
>         }
> ```
> 

I don't think the original reflection code had casts if Converter was missing, I had made sure all properties have proper converters.


> I although looked at the rest of the generated code and it is well formatted so nice work there.

I have spent a lot of time tweaking by hand the generated to look nice.

**Comment by inforithmics** [comment]
So from my side it is ok when this gets merged.

**Comment by wieslawsoltes** [comment]
Some stats about generated code:
```
      83 text files.
      83 unique files.                              
       0 files ignored.

github.com/AlDanial/cloc v 1.82  T=0.04 s (2032.1 files/s, 142224.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C#                              83            650             83           5076
-------------------------------------------------------------------------------
SUM:                            83            650             83           5076
```

**Comment by wieslawsoltes** [comment]
I have added `SvgTransformConverter` optimization:

### master branch

```
|                                                Method |     Mean |     Error |    StdDev |      Op/s | Rank |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------ |---------:|----------:|----------:|----------:|-----:|-------:|------:|------:|----------:|
|            SvgTransformConverter_ConvertFrom_Matrix_1 | 1.215 us | 0.0066 us | 0.0055 us | 823,166.9 |    3 | 0.0420 |     - |     - |     816 B |
|            SvgTransformConverter_ConvertFrom_Matrix_2 | 1.454 us | 0.0220 us | 0.0184 us | 687,809.2 |    4 | 0.0496 |     - |     - |     944 B |
|            SvgTransformConverter_ConvertFrom_Matrix_3 | 1.718 us | 0.0111 us | 0.0103 us | 582,070.5 |    6 | 0.0572 |     - |     - |    1072 B |
|            SvgTransformConverter_ConvertFrom_Matrix_4 | 1.546 us | 0.0037 us | 0.0031 us | 646,727.1 |    5 | 0.0534 |     - |     - |    1008 B |
|    SvgTransformConverter_ConvertFrom_Rotate_Translate | 1.120 us | 0.0078 us | 0.0069 us | 892,752.1 |    2 | 0.0534 |     - |     - |    1016 B |
|    SvgTransformConverter_ConvertFrom_Translate_Rotate | 1.104 us | 0.0062 us | 0.0055 us | 905,766.3 |    1 | 0.0515 |     - |     - |     992 B |
| SvgTransformConverter_ConvertFrom_Matrix_Rotate_Scale | 2.517 us | 0.0217 us | 0.0193 us | 397,304.6 |    7 | 0.0725 |     - |     - |    1400 B |

```

### Generators branch

```
|                                                Method |       Mean |    Error |  StdDev |        Op/s | Rank |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------ |-----------:|---------:|--------:|------------:|-----:|-------:|------:|------:|----------:|
|            SvgTransformConverter_ConvertFrom_Matrix_1 |   835.4 ns |  4.61 ns | 3.60 ns | 1,197,039.1 |    2 | 0.0105 |     - |     - |     200 B |
|            SvgTransformConverter_ConvertFrom_Matrix_2 |   946.3 ns | 10.86 ns | 9.07 ns | 1,056,754.7 |    3 | 0.0105 |     - |     - |     200 B |
|            SvgTransformConverter_ConvertFrom_Matrix_3 | 1,061.8 ns |  9.35 ns | 7.81 ns |   941,785.6 |    5 | 0.0095 |     - |     - |     200 B |
|            SvgTransformConverter_ConvertFrom_Matrix_4 |   994.5 ns |  9.14 ns | 8.10 ns | 1,005,520.3 |    4 | 0.0095 |     - |     - |     200 B |
|    SvgTransformConverter_ConvertFrom_Rotate_Translate |   609.0 ns |  9.34 ns | 7.80 ns | 1,641,989.4 |    1 | 0.0076 |     - |     - |     152 B |
|    SvgTransformConverter_ConvertFrom_Translate_Rotate |   612.9 ns | 10.64 ns | 9.95 ns | 1,631,627.9 |    1 | 0.0076 |     - |     - |     152 B |
| SvgTransformConverter_ConvertFrom_Matrix_Rotate_Scale | 1,738.7 ns | 10.92 ns | 9.68 ns |   575,137.4 |    6 | 0.0076 |     - |     - |     152 B |

**Comment by wieslawsoltes** [comment]
@mrbean-bremen This is ready to merge. Tested in SVG test app and tested if builds in Visual Studio, all works.

devenv_2021-01-07_19-08-40

**Pull request merged**
